### ISSUE TITLE
Retryer event publisher using retryOnResult(...)

### ISSUE THREAD

**Issue by JonJac** [resilience4j/resilience4j#413]
Hi,

I'm trying to use retryOnResult(...) to retry when a server returns status code 5xx from a http request. I've implemented the retryer as a decorator. The following snippet setup the retryer:

```
private void initialize(RetryConfiguration configuration) {
        RetryConfig config = RetryConfig.<HttpResponse>custom()
                .maxAttempts(configuration.getMaxRetryAttempts())
                .waitDuration(Duration.ofMillis(configuration.getWaitBetweenRetiresInMilliseconds()))
                .retryOnResult(response -> response.getStatusLine().getStatusCode() >= 500)
                .retryOnException(throwable -> API.Match(throwable).of(
                        API.Case($(Predicates.instanceOf(ConnectTimeoutException.class)), true),
                        API.Case($(), false)))
                .build();

        retryer = Retry.of("Http client", config);

        retryer.getEventPublisher()
                .onRetry(event -> logger.warn(event.toString()));
    }
```

The retryer is used like so:
```    @Override
    public HttpResponse execute(HttpUriRequest request, HttpContext context) throws IOException, ClientProtocolException {
        CheckedFunction0<HttpResponse> retryableSupplier = Retry.decorateCheckedSupplier(retryer, () -> httpClientDelegate.execute(request, context));
        return Try.of(retryableSupplier).get();
    }
```

My problem is that `retryer.getEventPublisher().onRetry(event -> logger.warn(event.toString()));` throws a nullpointer exception when a retry happens due to status code 5xx because `getLastThrowable()` returns null in the event's toString method:
```
public class RetryOnRetryEvent extends AbstractRetryEvent {

    private final Duration waitInterval;

    public RetryOnRetryEvent(String name, int numberOfAttempts, Throwable lastThrowable, long waitInterval) {
        super(name, numberOfAttempts, lastThrowable);
        this.waitInterval = Duration.ofMillis(waitInterval);
    }
    @Override
    public Type getEventType() {
        return Type.RETRY;
    }

    @Override
    public String toString() {
        return String.format("%s: Retry '%s', waiting %s until attempt #%d. Last attempt failed with exception %s",
                             getCreationTime(),
                             getName(),
                             waitInterval,
                             getNumberOfRetryAttempts(),
                             getLastThrowable().toString());
    }
}
```

Basically, the event is useless unless its triggered by an exception. If its trigged by retryOnResult(...), then there is no information as to why it has happend and toString() is unsafe to call. The resulting object is not accessible, so its not possible to create a custom error message.

Am I going about logging retries the wrong way or have i just ran into a limitation of the framework?

Otherwise, thanks for a great project :)

**Comment by RobWin** [comment]
Hi thanks for the kind words,

I think you found a bug in ´the usage of `RetryOnRetryEvent `
The [RetryEventPublisherTest](https://github.com/resilience4j/resilience4j/blob/master/resilience4j-retry/src/test/java/io/github/resilience4j/retry/RetryEventPublisherTest.java) is missing to test it.

Can you use `retry.getEventPublisher().onError` or `retry.getEventPublisher().onSuccess` instead?

**Comment by RobWin** [comment]
No wait. Which version are you using. I think it's fixed in version 0.14.1

**Comment by RobWin** [comment]
Anyway better use onSuccess or onError.

**Comment by JonJac** [comment]
I'm yet to find clear documentation on onSuccess and onError, but i specifically want to log why a retry was attempted. The do not trigger that to my understanding. Furthermore they have the same issue. with getLastThrowable().

**Comment by JonJac** [comment]
The nullpointer exception in toString() was fixed. The bigger design issue is still there: How am i supposed to log what triggered the retry, when it happens because of .retryOnResult(). There is no information in the event.

**Comment by RobWin** [comment]
True, you can only check if lastThrowable is null at the moment.

**Comment by RobWin** [comment]
Which information do you need to understand what triggered the retry?
Would you expect an exception like `RetryDueToResultException`

**Comment by JonJac** [comment]
The most pragmatic way would be that the event contained the result object. The .retryOnResult() method allows one to define predicates to be tested against the result object. **I should be able to tell exactly why a retry happened.** If i have the returned object, then i can run the predicate(s) against it and see what the retry cause is.

An alternative is that the user would provide a list of predicates to test against the result. The framework would then tell exactly which predicate caused the retry.

**Comment by JonJac** [comment]
Concretely, in my case of retrying on http calls, the resulting object would provide important state such as the status code. I could make a general predicate like .getStatusCode() >= 500 and then print the exact status code. Its very interesting information.

**Comment by RobWin** [comment]
Do you just want to log it with a logging framework? Or do you want to process the events and store information in another system?

**Comment by JonJac** [comment]
I just want to log it with a logging framework in this concrete case.

**Comment by Romeh** [comment]
you can not just log it inside the result predicate itself  ?

**Comment by RobWin** [comment]
Correct, that was also my idea.

**Comment by JonJac** [comment]
Good point. That would be viable here.

However, if that the intended way then i see little purpose for the event system. Would still think its neater if the returned object is available in events.

**Comment by RobWin** [comment]
The event system is used by use to connect monitoring systems (TSDB) like Prometheus, InfluxDB

**Comment by Romeh** [comment]
@RobWin  do we still need to keep that open ? i mean do we still need to do further actions ?
